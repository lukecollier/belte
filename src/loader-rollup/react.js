import * as rollup from 'rollup';

export async function loader(src, { encodeForFilename, encodeForVariableName }) {
  const inputOptions = {
    input: src,
    plugins: []
  }
  const bundle = await rollup.rollup(inputOptions);
  // logic here
  return {
    render: (src) => doARender(src, encode),
    plugins: () => [],
    dependencies: () => [
      'path/to/depOne.js',
      'path/to/depTwo.js'
    ] // might not need
  }
}
