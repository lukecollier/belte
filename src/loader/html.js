const svelte = require('svelte');
import * as R from 'ramda';
import { readFileSync } from 'fs';
import  pathUtil from 'path';
import { walk } from 'estree-walker';
const acorn = require('acorn');

export const render = (src, _) => {
  return readFileSync(src, 'utf8')
}
