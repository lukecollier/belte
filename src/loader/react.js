import ReactDOMServer from 'react-dom/server';
import React, { Component } from 'react';
import * as rollup from 'rollup';
import * as R from 'ramda';
import commonjs from 'rollup-plugin-commonjs';
import sucrase from 'rollup-plugin-sucrase';
import { readFileSync } from 'fs';
import * as pathUtil from 'path';
import isReact from 'is-react';

import { nameFromPath } from '../string.js';
import { walk } from 'estree-walker';
const acorn = require("acorn").Parser.extend(require("acorn-jsx")());

// dependency package now handles dependencies from an output file,
// client (renamed build) - now builds the client to esm standard preffered takes any source
// dependencies - done in core 
// render - same
// constructor - same
// otherDependencies - done in core 
// styles - out of scope for now

export async function asyncClient(src, encodeForVariableName) {
  const data = readFileSync(src, 'utf8');
  const dir = pathUtil.dirname(src);
  const componentsSrc = imports(data, dir);
  const components = {};
  componentsSrc.forEach(src => components[src] = encodeForVariableName(readFileSync(src, 'utf8')));
  const inputOptions = {
    input: src,
    external: (id) => componentsSrc.includes(id) || id === 'react',
    plugins: [ 
      sucrase({
        exclude: ['node_modules/**'],
        transforms: ['jsx']
      }),
      commonjs({include: 'node_modules/**'}) 
    ],
  }
  const outputOptions = {
    format: 'iife',
    name: nameFromPath(src),
    globals: Object.assign(components, { react: 'React' }),
    file: 'test.js',
  }
  const bundle = await rollup.rollup(inputOptions);
  const { output } = await bundle.generate(outputOptions);

  var result;
  for (const chunkOrAsset of output) {
    if (chunkOrAsset.isAsset) {
    } else { 
      result = chunkOrAsset.code;
    }
  }
  return result;
}

export const render = (src, attr) => {
	const Component = require(src);
	const comp = React.createElement(Component, attr);
	return ReactDOMServer.renderToString(comp)
}

export const constructor = (name, id, attr) => `ReactDOM.hydrate(React.createElement(${name}, ${JSON.stringify(attr)}), document.getElementById("${id}"));`;

export const client = (code, name, encode) => {
  const ast = recast.parse(code, {praser: acorn, sourceType: 'module'});
  recast.visit(ast, { 
    visitImportDeclaration: function(path) {
      path.prune();
      return false;
    },
    visitExportDefaultDeclaration: function (path) {
      path.replace(`return ${path.value.declaration.name};`);
      return false;
    }  
  });
  return `
    var ${name} = (function(${deps.join(',')}) {
      ${recast.print(ast).code}
    })(${deps.join(',')})
  `.trim();
}

export const imports = (code, dir) => {
  var deps = [];
  const ast = acorn.parse(code, {sourceType: 'module'});
  walk(ast, { enter: ( node, parent ) => {
    if (node.type === 'ImportDeclaration') {
      const path =  node.source.value.toString();
      const src = pathUtil.resolve(dir, path);
      if (src.endsWith('.jsx') && isReact.component(require(src).default)) {
        deps.push(src);
      }
    }
  }});
  return deps;
}

// loaders will no longer return a mother object but instead export the respective files to be included in the regular way. allows for a simpler dx.
export const loader = (src, encode) => {
  const data = readFileSync(src, 'utf8');
  const dir = pathUtil.dirname(src);
  const name = encode(data);

  return {
    render: R.partial(render, [src]),
    client: R.partial(client, [dir, data, encode]),
    styles: () => [],
    constructor: R.partial(constructor, [name]),
    dependencies: () => R.partial(imports, [data, dir]),  
    components: () => R.partial(imports, [data, dir]),
  }
};
