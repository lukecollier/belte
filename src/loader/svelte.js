

const svelte = require('svelte');
const acorn = require('acorn');
import { walk } from 'estree-walker';
import Hashids from 'hashids';
import * as R from 'ramda';

import fs from 'fs';
import pathUtil from 'path';

import { nameFromPath, filenameFromPath } from '../string.js';
import { encodeContent } from '../encoding.js';

const dependencies = (data, filePath) => {
  var deps = new Set();
  const dir = pathUtil.dirname(filePath);
	const options = {
		generate: 'dom',
		css: false,	
		hydratable: true,
    name: nameFromPath(filePath),
    filename: filenameFromPath(filePath),
		format: 'es'
	};
	const compiled = svelte.compile(data, options);
  const ast = acorn.parse(compiled.js.code, {sourceType: 'module'});
  walk(ast, {
    enter: ( node, parent ) => {
      if (node.type === 'ImportDeclaration') {
        const valuePath =  node.source.value.toString();
        if (valuePath.endsWith('.html')) {
          const src = pathUtil.resolve(dir, valuePath);
          const compData = fs.readFileSync(src, 'utf8');
          dependencies(compData, src).forEach((srcPath, alias) => { 
            deps.add(srcPath);
          });
          deps.add(valuePath);
        }
      }
    },
    leave: ( node, parent ) => {
    }
  });
  return deps;
}

export const compileClient = (src, encode) => {
  const compData = fs.readFileSync(src, 'utf8');
  const dir = pathUtil.dirname(src);
	const options = {
		generate: 'dom',
		css: false,	
		hydratable: true,
    name: encode(compData, 'Svelte'),
    filename: encode(compData, 'Svelte.'),
		format: 'iife',
    globals: (relPath) => {
      const resolved = pathUtil.resolve(dir, relPath);
      const data = fs.readFileSync(resolved, 'utf8');
      return encode(data, 'Svelte');
    }
	};
	const compiled = svelte.compile(compData, options);
	return compiled;
};

export const makeRender = (path) => {
  const src = pathUtil.resolve(process.cwd(), path);
  require('svelte/ssr/register');
  const component = require(src);
  return (attr) => component.render(attr).html;
}

export const resolveDependencies = (src, encode) => {
  const data = fs.readFileSync(src, 'utf8');
  var result = new Set();
  const deps = dependencies(data, src);
  const dir = pathUtil.dirname(src);
  for (const key of deps.keys()) {
    const resolvedPath = pathUtil.resolve(dir, key);
    const {js, css} = compileClient(resolvedPath, encode);
    result.add({js: js.code, css: css.code});
  };
  
  return Array.from(result); 
}

const constructor = (constructorName, instances) => 
  instances.map(instance => `new ${constructorName}({target:document.getElementById('${instance.id}'),hydrate:true,data:${JSON.stringify(instance.attr)}});`).join('');

export const loader = (src, instances = [{id: '', attr: {}}], 
  encode = R.partialRight(encodeContent, ['salt'])) => {

  const clientDepsCompiled = resolveDependencies(src, encode);
  const scriptsDeps = clientDepsCompiled.map(compile => compile.js);
  const stylesDeps = clientDepsCompiled.map(compile => compile.css)
    .filter(style => style !== null);
  const {js, css} = compileClient(src, encode); 
  const compData = fs.readFileSync(src, 'utf8');
  const initScript = constructor(encode(compData, 'Svelte'), instances);

  return {
    scripts: [
      ...scriptsDeps, 
      js.code, 
      initScript
    ],
    styles: (css.code !== null) ? [css.code, ...stylesDeps] : stylesDeps  
  };
};

export default loader;
