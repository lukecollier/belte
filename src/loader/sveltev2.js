const svelte = require('svelte');
import * as R from 'ramda';
import { readFileSync } from 'fs';
import  pathUtil from 'path';
import { walk } from 'estree-walker';
const acorn = require('acorn');

export const render = (src, attr) => {
  require('svelte/ssr/register');
  const component = require(src);
  return require(src).render(attr).html
}

export const client = (src, encode) => {
  const comp = readFileSync(src, 'utf8');
  const dir = pathUtil.dirname(src);
	const options = {
		generate: 'dom',
		css: false,	
		hydratable: true,
    name: 'Svelte' + encode(comp),
    filename: 'Svelte' + encode(comp) + '.js',
		format: 'iife',
    globals: (relPath) => {
      const data = readFileSync(pathUtil.resolve(dir, relPath), 'utf8');
      return 'Svelte' + encode(data);
    }
	};
	return svelte.compile(comp, options).js.code;
}

export const style = (src) => {
  const data = readFileSync(src, 'utf8');
	const options = {
		generate: 'dom',
		css: true,	
		format: 'cjs'
	};
	return svelte.compile(data, options).css.code;
}

export const constructor = (name, id, attr) => 
  `new ${"Svelte" + name}({target:document.getElementById('${id}'),hydrate:true,data:${JSON.stringify(attr)}});`;

const builtins = require("module").builtinModules;
export const imports = (code, dir) => {
  var deps = new Set();
  walk(acorn.parse(code, {sourceType: 'module'}), {
    enter: ( node, parent ) => {
      if (node.type === 'ImportDeclaration') {
        const source =  node.source.value.toString();
        if (source.endsWith('.html')) {
          deps.add(pathUtil.resolve(dir, source));
        } else if (!builtins.includes(source)) {
          const path = pathUtil.resolve(dir, source)
          deps.add(require.resolve(path));
        } else {
          throw Error('Module does not exist or is a node module');
        }
      }
    },
    leave: ( node, parent ) => {}
  });
  return Array.from(deps);
}

export const deps = (src) => {
  const data = readFileSync(src, 'utf8');
	const options = {
		generate: 'dom',
		css: false,	
		format: 'es'
	};
	const code = svelte.compile(data, options).js.code;
  return imports(code, pathUtil.dirname(src)).map(src => { 
    return [src, deps(src)].flat()
  }).flat();
}

export const loader = (src, encode) => {
  const data = readFileSync(src, 'utf8');
  const name = encode(data);

  const compDeps = () => deps(src).filter(filepath => filepath.endsWith('.html')); 
  const otherDeps = () => deps(src).filter(filepath => !filepath.endsWith('.html')); 

  return {
    render: R.partial(render, [src]),
    client: R.partial(client, [src, encode]),
    styles: R.partial(style, [src]),
    constructor: R.partial(constructor, [name]),
    dependencies: compDeps,  
    otherDependencies: otherDeps,  
  }
};

export default loader;
