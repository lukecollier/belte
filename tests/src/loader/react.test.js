import { loaderTest, loader } from '../../../src/loader/react';
import path from 'path';

test('loader can be loaded', () => {
  const result = loader(path.resolve(__dirname, '../../resource/react/StatefulComponent.jsx'), (_) => _);
  expect(result).toMatchSnapshot();
});

