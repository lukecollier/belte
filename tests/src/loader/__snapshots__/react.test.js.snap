// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader can be loaded 1`] = `
Object {
  "client": [Function],
  "components": [Function],
  "constructor": [Function],
  "dependencies": [Function],
  "render": [Function],
  "styles": [Function],
}
`;
